tests/testthat/test-dfm_group.R

```diff
@@ -56,3 +56,94 @@ test_that("dfm.character groups works (#794)", {
+        matrix(c(1,2,3,1, 1,0,0,0), byrow = TRUE, nrow = 2, 
+               dimnames = list(docs = c('doc1', 'doc2'), features = letters[1:4]))
+    )
+    expect_equal(
+        as.matrix(dfm_group(testdfm, groups = external_factor, fill = TRUE)),
+        matrix(c(0,0,0,0, 1,2,3,1, 1,0,0,0, 0,0,0,0), byrow = TRUE, nrow = 4, 
+               dimnames = list(docs = paste0("doc", 0:3), features = letters[1:4]))
+    )
+    # new documents in factor order
+    expect_equal(
+        as.matrix(dfm_group(testdfm, groups = factor(c(1, 1, 2), levels = 4:1), fill = TRUE)),
+        matrix(c(rep(0, 8), 1,0,0,0, 1,2,3,1), byrow = TRUE, nrow = 4,
+               dimnames = list(docs = 4:1, features = letters[1:4]))
+    )
+    # should this also be ordered? (here the expectation is that it is)
+    expect_equal(
+        as.matrix(dfm_group(testdfm, groups = factor(c(3, 3, 1), levels = 4:1), fill = FALSE)),
+        matrix(c(1,2,3,1, 1,0,0,0), byrow = TRUE, nrow = 4,
+               dimnames = list(docs = c(3, 1), features = letters[1:4]))
+    )
+})
+
+test_that("test dfm_group with non-factor grouping variable, with fill", {
+    grp <- c("D", "D", "A", "C")
+    corp <- corpus(c('a b c c', 'b c d', 'a', "b d d"),
+                   docvars = data.frame(grp = grp, stringsAsFactors = FALSE))
+    testdfm <- dfm(corp)
+    expect_equal(
+        as.matrix(dfm_group(testdfm, groups = "grp", fill = FALSE)),
+        matrix(c(1,0,0,0, 0,1,0,2, 1,2,3,1), byrow = TRUE, nrow = 3, 
+               dimnames = list(docs = c("A", "B", "D"), features = letters[1:4]))
+    )
+    expect_equal(
+        dfm_group(testdfm, groups = "grp", fill = FALSE),
+        dfm_group(testdfm, groups = "grp", fill = TRUE)
+    )
+    expect_equal(
+        dfm_group(testdfm, groups = grp, fill = FALSE),
+        dfm_group(testdfm, groups = grp, fill = TRUE)
+    )
+    # should this be sorted? I think it should be the same as when obtained from a docvar
+    expect_equal(
+        dfm_group(testdfm, groups = grp, fill = FALSE),
+        matrix(c(1,0,0,0, 0,1,0,2, 1,2,3,1), byrow = TRUE, nrow = 3, 
+               dimnames = list(docs = c("A", "B", "D"), features = letters[1:4]))
+    )
+    expect_equal(
+        dfm_group(testdfm, groups = grp, fill = FALSE),
+        dfm_group(testdfm, groups = "grp", fill = FALSE),
+    )
+})
+    
+test_that("test dfm_group with wrongly dimensioned groups variables", {
+    corp <- corpus(c('a b c c', 'b c d', 'a', "b d d"),
+                   docvars = data.frame(grp = grp, stringsAsFactors = FALSE))
+    testdfm <- dfm(corp)
+    expect_error(
+        dfm_group(testdfm, groups = c(1, 1, 2, 3, 3), fill = FALSE),
+        "groups must name docvars or provide data matching the documents in x"
+    )
+    expect_error(
+        dfm_group(testdfm, groups = c(1, 1, 2, 3, 3), fill = TRUE),
+        "groups must name docvars or provide data matching the documents in x"
+    )
+    expect_error(
```

Comment:
It should be an error to produce:
```r
dfm_group(testdfm, groups = c(1, 1, 2, 3, 4), fill = TRUE)
# Document-feature matrix of: 4 documents, 8 features (71.9% sparse).
# 4 x 8 sparse Matrix of class "dfmSparse"
#     features
# docs a B c D e BBB Aaaa cc
#    1 2 1 2 2 2   1    0  0
#    2 0 0 0 0 0   1    1  1
#    3 0 0 0 0 0   0    0  0
#    4 0 0 0 0 0   0    0  0
```
We should always check that the `groups` object is either a docvar name, or equal in length (or nrow) to the number of documents.